bpo-1812: Fix newline conversion when doctest.testfile loads from a package whose loader has a get_data method.

pdonis opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

[bpo-1812](https://bugs.python.org/issue1812): Fix newline conversion when doctest.testfile loads from a package whose loader has a get_data method.

This pull request fixes the newline conversion bug originally reported in issue 1812. When that issue was originally submitted, the open builtin did not default to universal newline mode; now it does, which makes the issue fix simpler, since the only code path that needs to be changed is the one in doctest._load_testfile where the file is loaded from a package whose loader has a get_data method. Commit b6ad8ee contains the change that implements the fix.

This pull request also adds a test that verifies that doctest.testfile does the newline conversion correctly when loading from a package whose loader has a get_data method. This test fails (a ValueError is raised for inconsistent whitespace) unless the change in commit b6ad8ee is applied. Commit dd02845 contains the changes that add the test; the test is added to the test_lineendings docstring along with supporting comments.

Also, the test_lineendings docstring is missing a test to check for correct handling of the old-style Mac newline \r. This pull request adds that case to the original test_lineendings docstring. Commit da35562 contains the changes that add this test and edit the surrounding comments accordingly.

Finally, this pull request contains commit 22094c8 that adds my name to the MISC/Acks file.


<!-- issue-number: [bpo-1812](https://bugs.python.org/issue1812) -->
https://bugs.python.org/issue1812
<!-- /issue-number -->

pdonis committed:
bpo-1812: Add old-style Mac newline test to test_doctest.test_lineendings.

The test_lineendings test in test_doctest did not include a test with
old-style Mac line endings (\r). This commit adds one, and edits the
introductory discussion in test_lineendings accordingly.

pdonis committed:
bpo-1812: Add test to test_doctest.test_lineendings that loads testfile from package with loader.get_data.

The previous test_doctest.test_lineendings did not cover the case where a testfile for doctest.testfile
is loaded from a package whose loader has a get_data method. This commit adds a test for that case.

pdonis committed:
Merge branch 'master' into fix-issue-1812

pdonis committed:
bpo-1812: Fix newline conversion when loading doctest testfile from package with loader.get_data.

This commit corrects the newline conversion that is done in doctest._load_testfile when the file
is being loaded from a package whose loader has a get_data method.

pdonis committed:
Add Peter Donis to Misc/ACKS.

pdonis committed:
Merge branch 'master' into fix-issue-1812

pdonis committed:
Merge branch 'master' into fix-issue-1812

pdonis committed:
Add MISC/NEWS.d blurb.

zware commented on Lib/test/test_doctest.py:
```diff
@@ -2687,6 +2720,38 @@ def test_lineendings(): r"""
+    >>> support.create_empty_file(os.path.join(pkg, "__init__.py"))
+    >>> fn = os.path.join(pkg, "doctest_testfile.txt")
+    >>> with open(fn, 'wb') as f:
+    ...     f.write(b'Test:\r\n\r\n  >>> x = 1 + 1\r\n\r\nDone.\r\nTest:\n\n  >>> x = 1 + 1\n\nDone.\nTest:\r\r  >>> x = 1 + 1\r\rDone.\r')
```
It would be nice to break up the long line here; it should be doable to split it up in some logical places and allow implicit concatenation to reconstruct it:

```suggestion
    ...     f.write(
    ...         b'Test:\r\n\r\n'
    ...         b'  >>> x = 1 + 1\r\n\r\n'
    ...         b'Done.\r\n'
    ...         b'Test:\n\n'
    ...         b'  >>> x = 1 + 1\n\n'
    ...         b'Done.\n'
    ...         b'Test:\r\r'
    ...         b'  >>> x = 1+ 1\r\r'
    ...         b'Done.\r'
    ...     )
```

zware commented on Lib/test/test_doctest.py:
```diff
@@ -2659,12 +2662,42 @@ def test_testfile(): r"""
+        sys.path_hooks[:] = self.path_hooks
+        sys.path_importer_cache.clear()
+        sys.modules.clear()
+        sys.modules.update(self.modules_before)
```
It would be better to do this in a context manager just to be very sure we don't miss the cleanup.  There are some helpers in `test.test_importlib.util` that may be helpful here, though I confess I don't understand the inner workings of `importlib` well enough to know exactly what we need here :).  On the other hand, perhaps we can just pass in a mock object for `package` and avoid messing with global import state at all.

pdonis replied:
I looked through `test.test_importlib.util` and didn't see anything that I thought I could use. I'm also not confident enough in my knowledge of how packages work to try to construct a mock package object that won't cause some other problem.

The context manager suggestion is a good one. It looks straightforward enough, I will test that and push an update if I can get it working.

pdonis replied:
The context manager turned out to be very simple; `contextlib.contextmanager` is my friend. :-)

There are a total of three new commits:

Commit efabd42 does the breakup of the long line (I just committed your suggested change, @zware).

Commit 1d56c05 fixes a typo in the breakup of the long line.

Commit e1f0a2f implements the context manager.

pdonis replied:
@zware Thanks for the kudos, btw!

brettcannon replied:
You want `test.test_importlib.util.import_state()`: https://github.com/python/cpython/blob/master/Lib/test/test_importlib/util.py#L183

zware replied:
@pdonis, did you see @brettcannon's message here?  I think there is some opportunity to clean up the clean up code a bit, but on the other hand that can be done in a follow-on PR.

zware reviewed (commented):
Overall, this is an incredibly clean patch, and you have the patience of a saint to have shepherded this along for 12 years!  I have a style nit and a cleanup suggestion, but this is probably acceptable as is.

Thanks for the patch, and hopefully we'll finally get this landed!

pdonis committed:
Update Lib/test/test_doctest.py

Break up long line to improve readability.

Co-Authored-By: Zachary Ware <zachary.ware@gmail.com>

pdonis committed:
Fix typo in commit efabd422.

pdonis committed:
bpo-1812: Wrap package hook test in context manager.

This commit wraps the doctest.testfile test with a package loader that
has a get_data method in a context manager to ensure proper removal
of the extra package hook.

ammaraskar commented on Lib/doctest.py:
```diff
@@ -221,7 +228,7 @@ def _load_testfile(filename, package, module_relative, encoding):
                 # get_data() opens files as 'rb', so one must do the equivalent
                 # conversion as universal newlines would do.
-                return file_contents.replace(os.linesep, '\n'), filename
+                return _newline_convert(file_contents), filename
```
Instead of trying to replicate the behavior of universal new lines, perhaps we can combine this and the `decode` routine above with something like:

```python
>>> file_contents = b"line1\r\nline2\nline3\rline4"
>>> io.TextIOWrapper(io.BytesIO(file_contents), encoding='utf8').read()
'line1\nline2\nline3\nline4'
```

pdonis replied:
I don't know enough about the internals of TextIOWrapper and BytesIO to know if this would risk breaking anything else. My intent was to make the minimal possible fix to the existing code, and since the existing code used the string replace operation, I just fixed that up to do the correct replacements.

I also don't know if the read operation on TextIOWrapper, or the BytesIO constructor, might throw different exceptions on malformed data than the simple decode and string replace operations would.

zware replied:
The concern about possibly changing exceptions is enough for me to want to stick with the simple fix here.

ammaraskar commented on Lib/test/test_doctest.py:
```diff
@@ -2659,12 +2663,52 @@ def test_testfile(): r"""
+        sys.modules.update(self.modules_before)
+
+
+@contextlib.contextmanager
```
I think some of this and the code above can be folded to use stuff from https://github.com/python/cpython/blob/master/Lib/test/test_importlib/fixtures.py like:

* https://github.com/python/cpython/blob/fd5116c0e77aec05f67fb24f10562ac920648035/Lib/test/test_importlib/fixtures.py#L71-L83

* https://github.com/python/cpython/blob/fd5116c0e77aec05f67fb24f10562ac920648035/Lib/test/test_importlib/fixtures.py#L50-L56

pdonis replied:
The OnSysPath fixture appears to be designed for a unittest test case, not doctest. A standalone version of the add_sys_path context manager might work if combined with install_finder; however, I notice that neither of those clears the path importer cache, and neither one removes the imported test module from sys.modules when the test is done. I don't know enough about the internals of the test framework to know if leaving out those last two items is OK; I put them in to make sure that everything that could possibly be mutated by this particular test was restored afterwards.

pdonis committed:
Merge branch 'master' into fix-issue-1812

brettcannon commented on Lib/test/test_doctest.py:
```diff
@@ -2659,12 +2663,52 @@ def test_testfile(): r"""
     >>> sys.argv = save_argv
 """
 
+class TestLoader(importlib.abc.MetaPathFinder, importlib.abc.ResourceLoader):
```
This is misnamed as this is implementing a finder and a loader, so call it an importer.

pdonis replied:
I'll push an update changing the name to TestImporter.

pdonis replied:
Commit bb836d7 does the name change.

pdonis committed:
Merge branch 'master' into fix-issue-1812

pdonis committed:
bpo-1812: Change TestLoader to TestImporter.

Change the name of the class used to test the case of a package with
a loader.get_data method from TestLoader to TestImporter.

zware reviewed (approved):
I wouldn't mind seeing the cleanup code cleaned up a bit in a separate PR, but as nobody has pointed out any actual *problems* with it as is, let's go ahead.
